p2p/client.go

```diff
@@ -2,111 +2,231 @@ package p2p
+const reAdvertisePeriod = 1 * time.Hour
+
+// Client is a P2P client, implemented with libp2p.
+//
+// Initially, client connects to predefined seed nodes (aka bootnodes, bootstrap nodes).
+// Those seed nodes serve Kademlia DHT protocol, and are agnostic to ORU chain. Using DHT
+// peer routing and discovery clients find other peers within ORU network.
 type Client struct {
 	conf    config.P2PConfig
+	chainID string
 	privKey crypto.PrivKey
 
+	ctx  context.Context
 	host host.Host
+	dht  *dht.IpfsDHT
+	disc *discovery.RoutingDiscovery
 
 	logger log.Logger
 }
 
-// NewClient creates new Client object
+// NewClient creates new Client object.
 //
 // Basic checks on parameters are done, and default parameters are provided for unset-configuration
-func NewClient(conf config.P2PConfig, privKey crypto.PrivKey, logger log.Logger) (*Client, error) {
+// TODO(tzdybal): consider passing entire config, not just P2P config, to reduce number of arguments
+// TODO(tzdybal): pass chainID
+func NewClient(ctx context.Context, conf config.P2PConfig, privKey crypto.PrivKey, logger log.Logger) (*Client, error) {
 	if privKey == nil {
 		return nil, ErrNoPrivKey
 	}
 	if conf.ListenAddress == "" {
-		// TODO(tzdybal): extract const
-		conf.ListenAddress = "/ip4/127.0.0.1/tcp/7676"
+		conf.ListenAddress = config.DefaultListenAddress
 	}
 	return &Client{
+		ctx:     ctx,
 		conf:    conf,
 		privKey: privKey,
 		logger:  logger,
 	}, nil
 }
 
+// Start establish Client's P2P connectivity.
+//
+// Following steps are taken:
+// 1. Setup libp2p host, start listening for incoming connections.
+// 2. Setup DHT, establish connection to seed nodes and initialize peer discovery.
+// 3. Use active peer discovery to look for peers from same ORU network.
 func (c *Client) Start() error {
-	c.logger.Debug("Starting P2P client")
+	c.logger.Debug("starting P2P client")
 	err := c.listen()
 	if err != nil {
 		return err
 	}
 
-	// start bootstrapping connections
-	err = c.bootstrap()
+	c.logger.Debug("setting up DHT")
+	err = c.setupDHT()
+	if err != nil {
+		return err
+	}
+
+	c.logger.Debug("setting up active peer discovery")
+	err = c.peerDiscovery()
 	if err != nil {
 		return err
 	}
 
 	return nil
 }
 
+// Close gently stops Client.
+func (c *Client) Close() error {
+	dhtErr := c.dht.Close()
+	if dhtErr != nil {
+		c.logger.Error("failed to close DHT", "error", dhtErr)
+	}
+	err := c.host.Close()
+	if err != nil {
+		c.logger.Error("failed to close P2P host", "error", err)
+		return err
+	}
+	return dhtErr
+}
+
 func (c *Client) listen() error {
-	// TODO(tzdybal): consider requiring listen address in multiaddress format
+	var err error
 	maddr, err := multiaddr.NewMultiaddr(c.conf.ListenAddress)
 	if err != nil {
 		return err
 	}
 
-	//TODO(tzdybal): think about per-client context
-	host, err := libp2p.New(context.Background(), libp2p.ListenAddrs(maddr), libp2p.Identity(c.privKey))
+	c.host, err = libp2p.New(c.ctx,
```

Comment:
Consider wrapping `Host` with [`RoutedHost`](https://github.com/libp2p/go-libp2p/blob/master/p2p/host/routed/routed.go#L32). It automatically resolves peers through Discovery if their addresses are not known.